feat: check media duration before download

jbsparrow opened the pull request:
Check media duration before downloading a file using [pyffmpeg](https://github.com/deuteronomy-works/pyffmpeg).

I will look into using `pyffmpeg` for our other ffmpeg functions but it will require a bit of work. The library does not bundle ffprobe, and instead just uses ffmpeg's log file to achieve the features that ffprobe would normally give you.

The library does not really provide functions for executing ffmpeg commands that we would use so I might open a PR to contribute or we can just use their binary with our existing logic.

Our ffprobe implementation is where it could be tedious.

jbsparrow committed:
Check Media Duration Before Download

Check media duration before downloading a file using `pyffmpeg`.

I will look into using `pyffmpeg` for our other ffmpeg functions but it will require a bit of work. The library does not bundle ffprobe, and instead just uses ffmpeg's log file to achieve the features that ffprobe would normally give you.

It seems to be made more so for simple usage by just running `FFmpeg().convert()`. Regardless, it does provide us with a bundled ffmpeg so it is worth switching the version that we use.

jbsparrow commented:
See https://github.com/jbsparrow/CyberDropDownloader/issues/762#issuecomment-3492908431 for info on Pyffmpeg and why I think we should not use it. For now it is okay but honestly, it might be better to just scrap this PR entirely and begin work on a custom ffprobe implementation. I do not want to work around the limitations of Pyffmpeg and I do not want to begin contributing a bunch of code to it. It doesn't actually include ffprobe, is slow, and their code is unfortunately quite a mess.

Let me know what you think @NTFSvolume.

NTFSvolume commented:
There's also this library: https://github.com/zackees/static_ffmpeg

It downloads the binaries at runtime on first use but we can write code that does that in a few dozens lines so i do not see much benefit

The actual benefit will be a package that includes the binaries in the wheel itself, so they are downloaded at the same time CDL is installed.

jbsparrow committed:
Revert "Check Media Duration Before Download"

This reverts commit addc819fe7b5b56119b850f456d4f8a0308cedea.

jbsparrow committed:
Remove dependence on videoprops

Make all our functions that relied on `videoprops` use our `ffmpeg` module.

Now videoprops only serves to ship ffprobe with CDL.

jbsparrow committed:
Use async for pre-download check

Use asynchronous probe for the pre-download file duration checks.

NTFSvolume commented on cyberdrop_dl/managers/client_manager.py:
```diff
@@ -245,12 +246,9 @@ def check_allowed_filetype(self, media_item: MediaItem) -> bool:
-            return True
-
-        return self.check_file_duration(media_item)
+        return await self.check_file_duration(media_item)
```
This method does nothing now so we can delete it and just call `self.check_file_duration` directly

NTFSvolume commented on cyberdrop_dl/managers/client_manager.py:
```diff
@@ -475,15 +473,20 @@ def check_file_duration(self, media_item: MediaItem) -> bool:
+
+            probe_path = media_item.complete_file if media_item.downloaded else media_item.url
+            probe_headers = self.download_client._get_download_headers(media_item.domain, media_item.referer)
+            properties: FFprobeResult = await probe(probe_path, headers=probe_headers)
```
Only parse headers for URLs
```suggestion
            if media_item.downloaded:
                probe_path = media_item.complete_file
                probe_headers = None 
            else:
                probe_path = media_item.url
                probe_headers = self.download_client._get_download_headers(media_item.domain, media_item.referer)
            properties: FFprobeResult = await probe(probe_path, headers=probe_headers)
```

NTFSvolume commented on cyberdrop_dl/managers/client_manager.py:
```diff
@@ -475,15 +473,20 @@ def check_file_duration(self, media_item: MediaItem) -> bool:
+                    return video.duration
+            if is_audio:
+                if audio := properties.audio:
+                    return audio.duration
```
```suggestion
            if is_video and (video := properties.video):
                return video.duration
            if is_audio and (audio := properties.audio):
                return audio.duration
```

NTFSvolume replied:
We may not even need the `is_video` and `is_audio` but just in case

NTFSvolume commented on cyberdrop_dl/managers/client_manager.py:
```diff
@@ -497,8 +500,9 @@ def get_duration() -> float | None:
             return True
 
-        duration: float = get_duration()  # type: ignore
+        duration: float = await get_duration()  # type: ignore
```
```suggestion
        duration: float | None = await get_duration()
```

NTFSvolume commented on cyberdrop_dl/utils/sorting.py:
```diff
@@ -134,10 +134,11 @@ async def sort_audio(self, file: Path, base_name: str) -> None:
-            duration = int(float(props.get("duration", 0))) or None
-            bitrate = int(float(props.get("bit_rate", 0))) or None
-            sample_rate = int(float(props.get("sample_rate", 0))) or None
+            properties: FFprobeResult = probe(file)
```
Why do we need the blocking probe? loop is already running here

jbsparrow replied:
You're right. I didn't realize that. In that case we do not need the synchronous probe function at all.

NTFSvolume commented on cyberdrop_dl/utils/sorting.py:
```diff
@@ -182,35 +183,29 @@ async def sort_video(self, file: Path, base_name: str) -> None:
+                width = video.width
+                height = video.height
+                if width and height:
+                    resolution = f"{width}x{height}"
```
```suggestion
                resolution = video.resolution
```

NTFSvolume reviewed (dismissed):
I can't think of a case in which we will need the blocking `probe`

Did you get any errors awaiting it?

jbsparrow committed:
Update cyberdrop_dl/managers/client_manager.py

Co-authored-by: NTFSvolume <172021377+NTFSvolume@users.noreply.github.com>

jbsparrow committed:
Update cyberdrop_dl/utils/sorting.py

Co-authored-by: NTFSvolume <172021377+NTFSvolume@users.noreply.github.com>

jbsparrow committed:
Update cyberdrop_dl/managers/client_manager.py

Co-authored-by: NTFSvolume <172021377+NTFSvolume@users.noreply.github.com>

jbsparrow committed:
Remove synchronous probe

Remove the synchronour probe function and implement requested changes.

NTFSvolume committed:
Merge branch 'dev' into pre-download-duration-check

NTFSvolume committed:
refactor: fix type errors

NTFSvolume committed:
perf: delay duration compute

NTFSvolume committed:
fix: handle sorting OSError

NTFSvolume committed:
fix: duration can not be float

NTFSvolume commented on cyberdrop_dl/managers/client_manager.py:
```diff
@@ -454,38 +447,51 @@ def check_file_duration(self, media_item: MediaItem) -> bool:
-                props: dict = get_audio_properties(str(media_item.complete_file))
-            return float(props.get("duration", 0)) or None
-
         duration_limits = self.manager.config.media_duration_limits
         min_video_duration: float = duration_limits.minimum_video_duration.total_seconds()
         max_video_duration: float = duration_limits.maximum_video_duration.total_seconds()
         min_audio_duration: float = duration_limits.minimum_audio_duration.total_seconds()
         max_audio_duration: float = duration_limits.maximum_audio_duration.total_seconds()
         video_duration_limits = min_video_duration, max_video_duration
         audio_duration_limits = min_audio_duration, max_audio_duration
```
These never change so we need to save them somewhere instead of computing them for every single file.

That's for another PR

NTFSvolume commented on cyberdrop_dl/managers/client_manager.py:
```diff
@@ -454,38 +447,51 @@ def check_file_duration(self, media_item: MediaItem) -> bool:
 
-        max_video_duration = max_video_duration or float("inf")
-        max_audio_duration = max_audio_duration or float("inf")
+        await self.manager.db_manager.history_table.add_duration(media_item.domain, media_item)
```
We will be adding the duration to the db twice for some items but that's fine for now.

NTFSvolume committed:
test: add ffprobe test

NTFSvolume committed:
refactor: better error message

NTFSvolume committed:
refactor: use truncated float for duration

Simplify formating

NTFSvolume committed:
test: relax tests for windows
